### ISSUE TITLE
fszmq won't find libzmq on macOS

### ISSUE THREAD

**Issue by mateuszlewko** [zeromq/fszmq#150]
I have installed zeromq using brew with this command:
`brew install zeromq --with-libpgm --with-libsodium`

But I get this error when creating Socket:
> Unhandled Exception:
System.DllNotFoundException: libzmq
  at (wrapper managed-to-native) fszmq.C:zmq_ctx_new ()
  at fszmq.Context..ctor (System.Object __) [0x00010] in <589095cd3d2ba9d0a7450383cd959058>:0
  at fszmq.Context..ctor () [0x00001] in <589095cd3d2ba9d0a7450383cd959058>:0

So I'm unable to use fszmq at all even though I've installed zeromq correctly.
I'm using zeromq 4.2.2 (I've also checked 4.0.8) and newest fszmq (12.2.2 available on nuget).

Do you know how to resolve this issue?

**Comment by pblasucci** [comment]
Thanks for raising this issue. I've run into this before and it's _usually_ just a case of Mono not knowing where to look for things. 

_(Aside: in some _rare_ cases, this gets reported erroneously when the actual error is a platform compatibility issue.)_

So there's not necessarily one single way to fix this. But I can suggest the following (in order of how I'd tackle it):

1. Ensure that homebrew has symlinked `libzmq.dylib` into `/usr/local/lib` _and_ that `/usr/local/lib` is in your `PATH` environment variable
2. Ensure that the full path to the folder containing `libzmmq.dylib` is in one of the environment variables mentioned in this article: http://www.mono-project.com/docs/advanced/pinvoke/
3. Copy `libzmq.dylib` from wherever homebrew has it _into_ the folder with your executable (Note: on some older versions of Mono this will have no effect)

Also, no matter how you play with pathing, it's important to specify a "DLL mapping" in either:

- your `<app>.config` file
- an `fszmq.dll.config` file, which is placed alongside `fszmq.dll` in your executable's folder (Note: you can find more info about this in the article linked above). There's an example of this mapping file here: https://github.com/zeromq/fszmq/blob/master/lib/zeromq/OSX/fszmq.dll.config

Please let me know if any of this helps or if you're still stuck. Thanks!

**Comment by mateuszlewko** [comment]
Thanks a lot for a response. 

So I've already added fszmq.dll.config file to my executable's directory and I've copied libzmq.dylib from hombrew's install folder (which was  /usr/local/Cellar/zeromq40/). Still the same error, except now it's `System.DllNotFoundException: libzmq.dylib` not just `libzmq`.

--Some time later--
In [this article](http://sushihangover.github.io/mono-unable-to-find-the-ldylib-native-library/) I've found how to check where mono is looking for libraries. It turns out that mono indeed found correct .dylib file but was unable load it:
> Did find:
        /usr/local/Cellar/zeromq/4.2.2/lib/libzmq.dylib: mach-o, but wrong architecture

So I've checked my library:

> file /usr/local/Cellar/zeromq/4.2.2/lib/libzmq.dylib
/usr/local/Cellar/zeromq/4.2.2/lib/libzmq.dylib: Mach-O 64-bit dynamically linked shared library x86_64

It should be fine as I'm on 64-bit mac.

Do you know what does ` but wrong architecture` means in this case? What should I do now?

Thanks!

**Comment by mateuszlewko** [comment]
Okay, so I've found the solution: 

The first problem was that mono couldn't find the library. We can solve this by for example placing .dylib file inside executable's folder or adding it's location to DYLD_LIBRARY_PATH.

Second problem was that mono couldn't load library. I've discovered this thanks to the article I mentioned above. The solution is to install `zeromq` using `brew` with `--universal` flag:
`brew install zeromq40 --universal`
It produces .dylib which contains library for both 32bit and 64bit:

> file /usr/local/Cellar/zeromq40/4.0.8/lib/libzmq.4.dylib
/usr/local/Cellar/zeromq40/4.0.8/lib/libzmq.4.dylib: Mach-O universal binary with 2 architectures: [x86_64: Mach-O 64-bit dynamically linked shared library x86_64] [i386: Mach-O dynamically linked shared library i386]
/usr/local/Cellar/zeromq40/4.0.8/lib/libzmq.4.dylib (for architecture x86_64):	Mach-O 64-bit dynamically linked shared library x86_64
/usr/local/Cellar/zeromq40/4.0.8/lib/libzmq.4.dylib (for architecture i386):	Mach-O dynamically linked shared library i386

Keep in mind that this flag applies for zeromq 4.0.x, for newest version I got this warning: 

> brew install zeromq --universal
Warning: zeromq: this formula has no --universal option so it will be ignored!
